binfinder.py

```diff
@@ -5,13 +5,99 @@
+    if start>=1.:
+        logger.error("pulsar is not in beam for any of the on-disk files. Ending...")
+        sys.exit(1)
+
+    #load presto module here because it uses python 2
+    commands.append('cd {0}'.format(pointing))
+    commands.append('echo "Folding on known pulsar {0}"'.format(pulsar))
+    commands.append('psrcat -e {0} > {0}.eph'.format(pulsar))
+    commands.append("sed -i '/UNITS           TCB/d' {0}.eph".format(pulsar))
+    commands.append("prepfold -o {0}_{2}_bins -noxwin -nosearch -runavg -noclip -timing {1}.eph"\
+                    " -nsub 256 -n {2} {3} -start {4} -end {5} 1*.fits"\
+                    .format(obsid, uplsar, nbins, mask, start, end))
+    commands.append('errorcode=$?')
+    commands.append('pulsar={}'.format(pulsar[1:]))
+    pulsar_bash_string = '${pulsar}'
+    #Some old ephems don't have the correct ra and dec formating and
+    #causes an error with -timing but not -psr
+    commands.append('if [ "$errorcode" != "0" ]; then')
+    commands.append('   echo "Folding using the -psr option"')
+    commands.append('   prepfold -o {0}_{2}_bins -noxwin -nosearch -runavg -noclip -psr {1}'\
+                    ' -nsub 256 -n {2} {3} -start {4} -end {5} 1*.fits'\
+                    .format(obsid, pulsar, nbins, mask, start, end))
+    commands.append('   pulsar={}'.format(pulsar))
+    commands.append('fi')
+    commands.append('rm {0}.eph'.format(pulsar))
+
+    return commands
+
+#----------------------------------------------------------------------
+def pulsar_beam_coverage(obsid, pulsar):
+    #returns the beginning and end time as a fraction that a pulsar is in the primary beam for the obsid files
+
+    #find the enter and exit times of pulsar normalized with the observing time
+    names_ra_dec = fpio.grab_source_alog(pulsar_list=[pulsar])
+    beam_source_data, _ = fpio.find_sources_in_obs([obsid], names_ra_dec)
+
+    enter_obs_norm = beam_source_data[obsid][0][1]
+    exit_obs_norm = beam_source_data[obsid][0][2]
+
+    #find the beginning and end time of the observation FILES you have on disk
+    files_beg, files_end = checks.find_beg_end(obsid)
+    files_duration = files_end - files_beg
+
+    #find how long the total observation is (because that's what enter and exit uses)
+    obs_beg, obs_end, obs_dur = file_maxmin.print_minmax(obsid)
+    obs_dur = obs_end-obs_beg
+     
+    #times the source enters and exits beam
+    time_enter = obs_beg + obs_dur*enter_obs_norm
+    time_exit = obs_beg + obs_dur*exit_obs_norm
+   
+    #normalised time the source enters/exits the beam in the files
+    enter_files = (time_enter-files_beg)/files_duration    
+    exit_files = (time_exit-files_beg)/files_duration
+
+    if enter_files<0.:
+        enter_files=0.
+    if exit_files>1.:
+        exit_files=1.
+    if enter_files>1.:
+        logger.warn("source {0} is not in the beam for the files on disk".format(pulsar))
```

Comment:
It also won't be in the beam if exit_files < 0

Reply:
yes but that should never happen because then the pulsar would never be considered in the first place

Reply:
Are you assuming that we'll always use the start of the data set?

Reply:
I'm assuming that the pulsar is considered a candidate for folding if find_pulsar_in_obs sends it to me. In which case, the pulsar must be in the primary beam at some point in the observation. If you pair this with the assumption that enter is always less than exit then exit must always be above zero, otherwise the pulsar is never in the beam for any of the observation, hence violating the first assumption.

Reply:
![70458364_493547384542115_93019546998276096_n](https://user-images.githubusercontent.com/18390846/64095850-9c237700-cd92-11e9-870b-bea1ade6a5fe.jpg)
Does this doodle help explain my point? You may only have data for the middle region (which is where my find_pusar_in_obs script may have seen the pulsar) but the data you have could be before or after it

Reply:
![20190902_153504](https://user-images.githubusercontent.com/34429506/64098009-4b168180-cd98-11e9-9481-a5b69a3d2242.jpg)

Okay after extensive doodling, i have decided that you are correct. If the files you have both begin and end after the pulsar leaves the beam in the observation, then you can get a negative exit case. I'll add this exception.